prompt.go

```diff
@@ -270,10 +270,16 @@ func (p *Prompt) feed(buf []byte) (shouldExit bool, exec *Exec, bufLeft []byte,
 		if p.handleASCIICodeBinding(b) {
 			return
 		}
+		i := 0
+		r, _ := utf8.DecodeRune(b[i:])
+		// AltGr key handling: expect ESC (0x1b) followed by Unicode character; skip ESC
+		if r == 0x1b {
+			i = 1
+			r, _ = utf8.DecodeRune(b[i:])
+		}
 		// Only insert printable characters
-		r, _ := utf8.DecodeRune(b)
 		if strconv.IsPrint(r) {
-			p.buf.InsertText(string(b), false, true)
+			p.buf.InsertText(string(b[i:]), false, true)
```

Comment:
Looks good!  It could be a little clearer to see some of the DecodeRune logic extracted into a fn and then some of the slice handling cleaned up along with it, perhaps like this:

```
		// Only insert printable characters
		r, _ := nextRune(b)
		if strconv.IsPrint(r) {
			p.buf.InsertText(string(r), false, true)
		}
...
func nextRune(b []byte) (r rune, size int) {
	r, size = utf8.DecodeRune(b)
	// Windows AltGr key handling: expect ESC (0x1b) followed by Unicode character; skip ESC
	if r == 0x1b {
		r, size = utf8.DecodeRune(b[1:])
	}
	return
}
```

I know the code here is general a bit of a mess though, so I don't want to be too critical of the working solution!

Reply:
I like that, but have a few concerns:

1. If we move the ESC skipping into a function like that, I worry that something else later may try to use that as a general purpose function and use it w/o understanding this behavior. Probably unlikely.
2. I'm not sure why the code was structured this way, but the test for a printable character is done with `r` and what is actually inserted into the input buffer is `b`. In theory at least multiple characters can follow the ESC, and I'm not sure if all of them would be properly encoded into `r`. I didn't want to change the logic too much because I didn't understand the ramifications.

If you don't think these are worth worrying about, then I'll create a separate function.

Reply:
1. I see your point.  How about then just giving it a more semantic name and/or a godoc comment to reflect it's special purpose?  Just trying to minimize the work being done directly in the switch case (which again is admittedly little like casting a stone in a glass house).
2. It reads to me like a bug if we are expecting `p.buf.InsertText(string(b[i:])` to handle the case where there might be more than one rune in `b[I:]`, since the `if` clause only checks that `r` is printable, so I feel like there's an issue one way or another.  If you look through the layers of input reading code that results in the `[]byte` passed to feed, there's nothing to indicate that it can't be more than one rune, so yours is not an unfounded concern.  Other parts of the feed code, though, clearly expect only a single rune entry in the byte slice in many cases, (see the behavior of `GetKey(b)` where if any of the special control chars were to come through as a part of a larger byte slice, then these special cases would be skipped).  I'm not 100% confident in this, but I believe that because the terminal is set to raw mode in UNIX and relies on `ReadRune` under the hood for Windows then we do in practice only ever see individual key strokes flowing through into feed.  In any case, we could leave the `string(b[i:])` approach, though then it would seem prudent to also step through each rune of `b` and check for printability before rendering into the final string.  The signature of `nextRune` could be updated to also return the stringified value of `r` to help encapsulate the `i` variable.

I don't feel strongly about any of it, so do what feels right and go ahead and merge it :)

Reply:
I realize the lack of checking the entire byte slice for printability is not your bug, just incidentally in the same section of the code.  Thinking a minute more about it, that printability test may arguably belong in `p.buf.InsertText` instead anyway.